network/p2p/acp118/aggregator.go

```diff
@@ -0,0 +1,294 @@
+// Copyright (C) 2019-2024, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+package acp118
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sync"
+
+	"github.com/ava-labs/avalanchego/utils/math"
+	"go.uber.org/zap"
+	"google.golang.org/protobuf/proto"
+
+	"github.com/ava-labs/avalanchego/ids"
+	"github.com/ava-labs/avalanchego/network/p2p"
+	"github.com/ava-labs/avalanchego/proto/pb/sdk"
+	"github.com/ava-labs/avalanchego/utils/crypto/bls"
+	"github.com/ava-labs/avalanchego/utils/logging"
+	"github.com/ava-labs/avalanchego/utils/set"
+	"github.com/ava-labs/avalanchego/vms/platformvm/warp"
+)
+
+var (
+	// ErrDuplicateValidator is returned if the provided validator set contains
+	// duplicate validators
+	ErrDuplicateValidator = errors.New("duplicate validator")
+	// ErrFailedAggregation is returned if it's not possible for us to
+	// generate a signature
+	ErrFailedAggregation = errors.New("failed aggregation")
+)
+
+type Validator struct {
+	NodeID    ids.NodeID
+	PublicKey *bls.PublicKey
+	Weight    uint64
+}
+
+type indexedValidator struct {
+	Validator
+	Index int
+}
+
+type result struct {
+	Signature *warp.BitSetSignature
+	Err       error
+}
+
+// NewSignatureAggregator returns an instance of SignatureAggregator
+func NewSignatureAggregator(
+	log logging.Logger,
+	client *p2p.Client,
+	maxPending int,
+) *SignatureAggregator {
+	return &SignatureAggregator{
+		log:        log,
+		client:     client,
+		maxPending: maxPending,
+	}
+}
+
+// SignatureAggregator aggregates validator signatures for warp messages
+type SignatureAggregator struct {
+	log        logging.Logger
+	client     *p2p.Client
+	maxPending int
+}
+
+// AggregateSignatures blocks until quorumNum/quorumDen signatures from
+// validators are requested to be aggregated into a warp message.
+func (s *SignatureAggregator) AggregateSignatures(
+	ctx context.Context,
+	message *warp.UnsignedMessage,
+	justification []byte,
+	validators []Validator,
+	quorumNumMin uint64,
+	quorumDenMin uint64,
+	quorumNumMax uint64,
+	quorumDenMax uint64,
+) (*warp.Message, error) {
```

Comment:
So the inclusion of these lower + upper bound parameters was to support DSMR's use-case where we want to meet a minimum threshold of signatures, but want to ideally replicate to more if possible. Another alternative to having these parameters here is to have a signature like so:

```suggestion
func (s *SignatureAggregator) AggregateSignatures(
	ctx context.Context,
	message *warp.Message,
	justification []byte,
	validators []Validator,
	quorumNum uint64,
	quorumDen uint64,
) (*warp.Message, error) {
```

If we pass in a signed message we can request signatures that aren't already in the bitset (will require a refactor to expose the concrete type of the `BitSetSignature` type on `warp.Message`). That way the caller could cancel the context on a timeout and decide if it wants to continue aggregation or not.

Reply:
The [signature-aggregator](https://github.com/ava-labs/awm-relayer/tree/main/signature-aggregator) service implementation caches signatures per Warp message ID, indexed by the signer's public keys, allowing us to reconstruct signed messages when the canonical validator set changes in certain cases, without having to request them again. The proposed alternative would work well for that use case, in addition to allowing the caller to cancel as already described.

Reply:
Related - for the first call to `AggregateSignatures`, is the idea to pass a `warp.Message` with an default initialized (or nil) `BitSetSignature`?

Reply:
Yeah, exaxctly. The normal usage would be to use a warp message with an empty signature.